### PR TITLE
[Wallet] Redesign of "Chose recipient" and "Enter amount" screens

### LINKED ISSUE #3911: New Design for Send Part 1 (till the TX review)

### PR DESCRIPTION
**Pull request by i1skn**
### Description

Implement choose Recipient screen and Amount enter screen for Send and Request flows. Search input is excluded for now and will updated later

### Tested

Android emulator and iOS simulator

### Related issues

- Fixes #3911

### Backwards compatibility

Yes

### Screenshots

Screenshot_1590434295
Screenshot_1590514745
Simulator Screen Shot - iPhone 11 Pro - 2020-05-27 at 02 16 55
Simulator Screen Shot - iPhone 11 Pro - 2020-05-27 at 02 17 26

### CODE REVIEW

File: packages/mobile/src/send/SendAmount.tsx [hunk 1]
```diff
@@ -1,254 +1,217 @@
...
-  fetchDollarBalance: typeof fetchDollarBalance
-  showMessage: typeof showMessage
-  showError: typeof showError
-  hideAlert: typeof hideAlert
-  fetchAddressesAndValidate: typeof fetchAddressesAndValidate
-}
+  React.useEffect(() => {
+    dispatch(fetchDollarBalance())
+    if (recipient.e164PhoneNumber) {
+      dispatch(fetchAddressesAndValidate(recipient.e164PhoneNumber))
```
Comment (i1skn) [inline packages/mobile/src/send/SendAmount.tsx]: This leads to "Enter PIN" screen appearing right when user is shown "Enter Amount" screen. This is not the best user experience, but I do not know how to solve it. @jmrossy is that something related to PGPNP, right?

File: packages/mobile/src/account/DollarEducation.tsx [hunk 2]
```diff
@@ -20,7 +20,7 @@ export class DollarEducation extends React.Component<Props> {
   goToSend = () => {
     this.props.setEducationCompleted()
     CeloAnalytics.track(CustomEventNames.send_dollar_nux)
-    navigate(Screens.Send)
+    navigate(Screens.Send, {})
```
Comment (aaronmgdr) [inline packages/mobile/src/account/DollarEducation.tsx]: I think if params are never needed it can be re typed to accept undefined so you dont need empty object

@cmcewen  is that right?

File: packages/mobile/src/navigator/Headers.v2.tsx [hunk 3]
```diff
@@ -77,6 +89,83 @@ export const emptyHeader: StackNavigationOptions = {
...
+  route: RouteProp<StackParamList, Screens.ExchangeTradeScreen>
+}) => {
+  const { makerToken } = route.params?.makerTokenDisplay
+  const title =
+    makerToken === CURRENCY_ENUM.DOLLAR
+      ? i18n.t('exchangeFlow9:buyGold')
+      : i18n.t('exchangeFlow9:sellGold')
+  return {
+    ...headerWithCancelButton,
+    headerLeft: () => <CancelButton style={{ color: colors.dark }} />,
```
Comment (aaronmgdr) [inline packages/mobile/src/navigator/Headers.v2.tsx]: would be nice to uninline this style

File: packages/mobile/src/account/DollarEducation.tsx (hunk 2 above)
Comment (i1skn) [reply packages/mobile/src/account/DollarEducation.tsx]: My bad, will add `| undefined` in `navigator/types.tsx`

File: packages/mobile/locales/en-US/sendFlow7.json [hunk 4]
```diff
@@ -105,5 +105,6 @@
     "body1": "Your contact can find their account number (also known as an address) at the top of the Settings menu.",
     "body2": "Account numbers begin with “0x” and are 42 characters long. They appear directly beneath the profile photo."
   },
-  "transferAddressChanged": "This contact changed their account number since the transaction. This payment was sent to:"
+  "transferAddressChanged": "This contact changed their account number since the transaction. This payment was sent to:",
+  "review": "Review"
```
Comment (jmrossy) [inline packages/mobile/locales/en-US/sendFlow7.json]: Consider putting one-word strings like this in `global.json` for easier use elsewhere

File: packages/mobile/src/navigator/Headers.v2.tsx [hunk 5]
```diff
@@ -77,6 +92,83 @@ export const emptyHeader: StackNavigationOptions = {
...
+  route: RouteProp<StackParamList, Screens.ExchangeTradeScreen>
+}) => {
+  const { makerToken } = route.params?.makerTokenDisplay
+  const title =
+    makerToken === CURRENCY_ENUM.DOLLAR
+      ? i18n.t('exchangeFlow9:buyGold')
+      : i18n.t('exchangeFlow9:sellGold')
+  return {
+    ...headerWithCancelButton,
+    headerLeft: () => <CancelButton style={styles.cancelBtn} />,
```
Comment (jmrossy) [inline packages/mobile/src/navigator/Headers.v2.tsx]: Nit, should cancelBtn just be the default style of this component?

File: packages/mobile/src/navigator/Headers.v2.tsx [hunk 6]
```diff
@@ -77,6 +92,83 @@ export const emptyHeader: StackNavigationOptions = {
...
+        testID="EditButton"
+        onPress={navigateBack}
+        titleStyle={{ color: colors.goldDark }}
+      />
+    ),
+    headerTitle: () => <HeaderTitleWithBalance title={title} token={makerToken} />,
+  }
+}
+
+export const sendOptions = ({ route }: { route: RouteProp<StackParamList, Screens.Send> }) => {
```
Comment (jmrossy) [inline packages/mobile/src/navigator/Headers.v2.tsx]: I think putting these in this file made sense when they were shared. But if they're going to be specific to one screen like these new ones, it'd be better to colocate it with the component. WDYT?

File: packages/mobile/src/send/Send.tsx [hunk 7]
```diff
@@ -187,6 +183,7 @@ class Send extends React.Component<Props, State> {
 
   onSelectRecipient = (recipient: Recipient) => {
     this.props.hideAlert()
+    const isRequest = this.props.route.params?.isRequest ?? false
```
Comment (jmrossy) [inline packages/mobile/src/send/Send.tsx:186]: Nice use of the `?` operators here ;)

File: packages/react-components/icons/Times.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import * as React from 'react'
+import Svg, { Path } from 'svgs'
+
+export interface Props {
+  height?: number
+  color?: string
+  strokeWidth?: number
+}
+
+function Times({ color, height, strokeWidth }: Props) {
```
Comment (jmrossy) [inline packages/react-components/icons/Times.tsx:11]: Same comment I left on Connor's last PR: I think icons should be PureComponents or React.Memo'ed

File: packages/mobile/src/send/SendAmount.tsx [hunk 9]
```diff
@@ -1,254 +1,217 @@
...
-)
-const CommentInput = withTextInputLabeling<TextInputProps>(TextInput)
+type RouteProps = StackScreenProps<StackParamList, Screens.SendAmount>
+type Props = RouteProps
 
-interface State {
-  amount: string
-  reason: string
-}
+const { decimalSeparator } = getNumberFormatSettings()
```
Comment (jmrossy) [inline packages/mobile/src/send/SendAmount.tsx:66]: Does this change if a user changes their language via settings?

File: packages/mobile/src/send/SendAmount.tsx (hunk 1 above)
Comment (jmrossy) [reply packages/mobile/src/send/SendAmount.tsx]: Yes, this is unfortunately necessary atm because PGPNP uses the wallet key to sign an auth token. We can explore alternatives in the future for using the DEK instead. @nityas and I talked about this already so she should be aware. It's essentially just bumping the pin up here instead of after review.

File: packages/mobile/src/send/SendAmount.tsx [hunk 10]
```diff
@@ -1,254 +1,217 @@
...
 
-interface DispatchProps {
-  fetchDollarBalance: typeof fetchDollarBalance
-  showMessage: typeof showMessage
-  showError: typeof showError
-  hideAlert: typeof hideAlert
-  fetchAddressesAndValidate: typeof fetchAddressesAndValidate
-}
+  React.useEffect(() => {
+    dispatch(fetchDollarBalance())
```
Comment (jmrossy) [inline packages/mobile/src/send/SendAmount.tsx:75]: do we want to do this each time?

File: packages/mobile/src/send/SendAmount.tsx [hunk 11]
```diff
@@ -1,254 +1,217 @@
...
+    if (recipient.e164PhoneNumber) {
```
Comment (jmrossy) [inline packages/mobile/src/send/SendAmount.tsx]: I believe there's a bug here, possibly from @tarikbellamine's changes for SecureSend.
We used to check if the recipient type was address or qrcode and skip the lookup in those cases, but that check seems to be gone:
https://github.com/celo-org/celo-monorepo/blob/6e6a9e5105796286064c4c7a9f903137d7d674d3/packages/mobile/src/send/SendAmount.tsx#L163

Comment (jmrossy) [reply packages/mobile/src/send/SendAmount.tsx]: @tarikbellamine do we handle this case somewhere else now?

**Review by jmrossy — COMMENTED** [review]
Mostly looks good! Some questions below

File: packages/mobile/src/send/SendAmount.tsx (hunk 11 above)
Comment (tarikbellamine) [reply packages/mobile/src/send/SendAmount.tsx]: My overarching reasoning was that we should fetch the latest mapping whenever we can as it's better to be safe that sorry with SecureSend.
- The false positives will at most require 2 or 3 more clicks and an extra 10-20 seconds to go through SecureSend as they already have their QR code or address ready to verify. Also worth noting that recipients caught here will have to go through SecureSend eventually anyways, we are just pulling the process forward to a transaction that may have not needed it.
- A false negative is a loss of funds. Admittedly, I don't know enough about how we assign recipient types to say how likely I think a false negative would be.

Given the uncertainty and the relatively low cost of playing it safe, that's the direction I went in. What do you think I could be missing?

File: packages/mobile/src/send/SendAmount.tsx (hunk 9 above)
Comment (i1skn) [reply packages/mobile/src/send/SendAmount.tsx:66]: Yes, but the app needs to be restarted. Essentially, even if it was not restarted - it would show previous `decimalSeparator`, but would be consistent around the app.

File: packages/mobile/src/send/SendAmount.tsx (hunk 10 above)
Comment (i1skn) [reply packages/mobile/src/send/SendAmount.tsx:75]: No, only once, hence I set the second parameter to `[]`

File: packages/mobile/src/navigator/Headers.v2.tsx (hunk 6 above)
Comment (i1skn) [reply packages/mobile/src/navigator/Headers.v2.tsx]: Yeah, I've been back and forth with this :) First did it in place, then decided to move them to `Headers`. But now, when you put this way I see it makes sense to have them in-place 👍

File: packages/react-components/icons/Times.tsx (hunk 8 above)
Comment (i1skn) [reply packages/react-components/icons/Times.tsx:11]: Thanks!

File: packages/mobile/src/fees/selectors.ts [hunk 12]
```diff
@@ -29,19 +29,21 @@ export const getReclaimEscrowFeeEstimateDollars = feeEstimateDollarsSelectorFact
   getReclaimEscrowFeeEstimateInWei
 )
 
-export const getFeeEstimateDollars = (state: RootState, feeType: FeeType | null) => {
-  if (feeType === null) {
-    return undefined
-  }
+export const getFeeEstimateDollars = (feeType: FeeType | null) => {
```
Comment (aaronmgdr) [inline packages/mobile/src/fees/selectors.ts]: while youre changing this anyway want to make it into a named function? 

Thats the guideline we have right? standalone functions named. class methods as arrows?

File: packages/mobile/src/navigator/types.tsx [hunk 13]
```diff
@@ -112,9 +112,14 @@ export type StackParamList = {
   [Screens.RegulatoryTerms]: undefined
   [Screens.Security]: undefined
   [Screens.SelectLocalCurrency]: undefined
-  [Screens.Send]: undefined
+  [Screens.Send]:
+    | {
```
Comment (aaronmgdr) [inline packages/mobile/src/navigator/types.tsx:116]: what does the pipe before the type do? between them its like an OR but before?

Comment (i1skn) [reply packages/mobile/src/navigator/types.tsx:116]: Yeah, Prettier adds it automatically. It is convenient for reading. 
```
...:
  | a
  | b
```
instead of
```
...: a 
  | b
```

File: packages/mobile/src/send/SendAmount.tsx (hunk 10 above)
Comment (jmrossy) [reply packages/mobile/src/send/SendAmount.tsx:75]: Sorry, I should have been more clear. Do we want to do this every time we load the SendAmount screen? Now that I think about it more I guess it's okay.

File: packages/mobile/src/send/SendAmount.tsx (hunk 11 above)
Comment (jmrossy) [reply packages/mobile/src/send/SendAmount.tsx]: That's all reasonable but we may still want to skip for these cases.
In the `address` case, I don't think we'll have e164Number set so we're probably fine either way.
In the QRCode case, my preference is that QR codes *always* work seamlessly. If we force a lookup in that case, we're vulnerable to a whole bunch of possible issues like:
1. PGPNP being down/unavailable
2. User being out of PGPNP lookups and having to buy more first
3. Recipient has incorrect/outdated data on-chain

QR code should take precedence over the on-chain mappings IMO because the user know who they're interacting with.

Comment (tarikbellamine) [reply packages/mobile/src/send/SendAmount.tsx]: I see. I wasn't aware of the costs associated with doing a lookup and agree that it is preferable to have one method (QR scanning) that always works, no matter what.

In that case, I am good with reverting back to the old conditional logic. From a process perspective, does it make more sense for me to open up a quick PR or let @i1skn revert that one line?

Comment (i1skn) [reply packages/mobile/src/send/SendAmount.tsx]: or just propose a change right in this PR using 

Screenshot 2020-05-27 at 20 19 23

Comment (tarikbellamine) [reply packages/mobile/src/send/SendAmount.tsx]: Thanks for the heads up on that feature. Looking at it a bit more, I think it may make more sense to submit a quick PR because it requires importing `Recipient` type and I'd feel better if I could also run the tests with the augmented logic to make sure there aren't downstream issues.

Sorry for the inconvenience!

Comment (tarikbellamine) [reply packages/mobile/src/send/SendAmount.tsx]: FYI it's been [merged into master](https://github.com/celo-org/celo-monorepo/pull/3915)

File: packages/mobile/src/recipients/RecipientItem.tsx [hunk 14]
```diff
@@ -24,55 +26,50 @@ class RecipientItem extends React.PureComponent<Props> {
     const { recipient } = this.props
 
     return (
-      <TouchableHighlight onPress={this.onPress} underlayColor={colors.altDarkBg}>
-        <View style={style.row}>
+      <Touchable onPress={this.onPress}>
```
Comment (jeanregisser) [inline packages/mobile/src/recipients/RecipientItem.tsx:29]: Nothing to do now but I think we should have a `TouchableItem` component to use in scrollable list of items that would use `TouchableHighlight` on iOS and something based on `TouchableNativeFeedback` on Android, to respect platform conventions.

`Touchable` alone currently uses `TouchableOpacity` on iOS and `TouchableNativeFeedback` on Android.

File: packages/react-components/icons/Times.tsx (hunk 8 above)
Comment (jeanregisser) [inline packages/react-components/icons/Times.tsx:11]: nit: Would "Close" be a more descriptive name for this icon?

File: packages/mobile/src/send/SendAmount.tsx [hunk 15]
```diff
@@ -1,254 +1,222 @@
...
 
-interface State {
-  amount: string
-  reason: string
-}
+const { decimalSeparator } = getNumberFormatSettings()
 
-type OwnProps = StackScreenProps<StackParamList, Screens.SendAmount>
-type Props = StateProps & DispatchProps & OwnProps & WithTranslation
+function SendAmount(props: Props) {
```
Comment (jeanregisser) [inline packages/mobile/src/send/SendAmount.tsx:68]: Nice refactor into a functional component 💪

**Review by jeanregisser — APPROVED** [review]
Awesome!! 👍

File: packages/react-components/icons/Times.tsx (hunk 8 above)
Comment (i1skn) [reply packages/react-components/icons/Times.tsx:11]: Usually I see this icon under "times" name - https://fontawesome.com/icons?d=gallery&q=times
